src/libstd/sys/windows/stdio.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use prelude::v1::*;
+use io::prelude::*;
+
+use io::{self, Cursor};
+use libc;
+use str;
+use sys::c;
+use sys::handle::Handle;
+use sys::cvt;
+
+struct NoClose(Option<Handle>);
+
+pub struct Stdin {
+    handle: NoClose,
+    utf8: io::Cursor<Vec<u8>>,
+}
+pub struct Stdout(NoClose);
+pub struct Stderr(NoClose);
+
+fn get(handle: libc::DWORD) -> io::Result<NoClose> {
+    let handle = unsafe { c::GetStdHandle(handle) };
+    if handle == libc::INVALID_HANDLE_VALUE {
+        Err(io::Error::last_os_error())
+    } else if handle.is_null() {
+        Err(io::Error::new(io::ErrorKind::Other,
+                           "no stdio handle available for this process", None))
+    } else {
+        Ok(NoClose::new(handle))
+    }
+}
+
+fn write(handle: libc::HANDLE, data: &[u8]) -> io::Result<usize> {
+    let utf16 = match str::from_utf8(buf).ok() {
+        Some(utf8) => utf8.utf16_units().collect::<Vec<u16>>(),
+        None => return Err(invalid_encoding()),
+    };
+    let mut written: DWORD = 0;
+    try!(cvt(unsafe {
+        WriteConsoleW(handle,
+                      utf16.as_ptr() as LPCVOID,
+                      utf16.len() as u32,
+                      &mut written as LPDWORD,
+                      ptr::null_mut())
+    }));
+
+    // FIXME if this only partially writes the utf16 buffer then we need to
+    //       figure out how many bytes of `data` were actually written
+    assert_eq!(written as usize, utf16.len());
+    Ok(data.len())
+}
+
+impl Stdin {
+    pub fn new() -> Stdin {
+        Stdin {
+            handle: get(c::STD_INPUT_HANDLE).unwrap(),
+            buf: Cursor::new(Vec::new()),
+        }
+    }
+
+    pub fn read(&self, buf: &mut [u8]) -> io::Result<usize> {
+        // Read more if the buffer is empty
+        if self.utf8.position() as usize == self.utf8.get_ref().len() {
+            let mut utf16: Vec<u16> = repeat(0u16).take(0x1000).collect();
```

Comment:
> The read method in this case does not give us &mut [u16] (which is what ReadConsoleW wants) and the only method of going from UCS-2 to &[u8] that is well defined (currently at least) is if the UCS-2 is valid UTF-16.

I think that `StdinRaw::read` should abstract over this difference, not `Stdin::read`. Currently one basically has no way to avoid having buffering and encoding conversions on Stdin without using `libc`, while, again the `sys` APIs were supposed to mirror the platform as closely as possible AFAIR. 

Implementation of this function also makes us lie to the users of `stdin_raw` function and we cannot avoid two layers of buffering in the buffered Stdin Reader either.